// happy-dom/cjs/index.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FocusEvent = exports.FileReader = exports.FileList = exports.File = exports.EventTarget = exports.EventPhaseEnum = exports.Event = exports.ErrorEvent = exports.Element = exports.DocumentType = exports.DocumentFragment = exports.Document = exports.DetachedBrowserPage = exports.DetachedBrowserFrame = exports.DetachedBrowserContext = exports.DetachedBrowser = exports.DataTransferItemList = exports.DataTransferItem = exports.DataTransfer = exports.DOMRect = exports.DOMParser = exports.DOMException = exports.CustomEvent = exports.CustomElementRegistry = exports.Comment = exports.ClipboardItem = exports.ClipboardEvent = exports.Clipboard = exports.CSSSupportsRule = exports.CSSStyleSheet = exports.CSSStyleRule = exports.CSSStyleDeclaration = exports.CSSRule = exports.CSSMediaRule = exports.CSSKeyframesRule = exports.CSSKeyframeRule = exports.CSSFontFaceRule = exports.CSSContainerRule = exports.BrowserWindow = exports.BrowserPage = exports.BrowserNavigationCrossOriginPolicyEnum = exports.BrowserFrame = exports.BrowserErrorCaptureEnum = exports.BrowserContext = exports.Browser = exports.Blob = exports.Attr = exports.AnimationEvent = exports.AbortSignal = exports.AbortController = void 0;
exports.HTMLFieldSetElement = exports.HTMLDirectoryElement = exports.HTMLDetailsElement = exports.HTMLDataListElement = exports.HTMLDataElement = exports.HTMLCanvasElement = exports.HTMLBRElement = exports.HTMLAreaElement = exports.HTMLDivElement = exports.HTMLDListElement = exports.HTMLMenuElement = exports.HTMLLIElement = exports.HTMLOListElement = exports.HTMLUListElement = exports.HTMLPreElement = exports.HTMLHRElement = exports.HTMLParagraphElement = exports.HTMLHeadingElement = exports.HTMLBodyElement = exports.HTMLHeadElement = exports.HTMLTitleElement = exports.HTMLVideoElement = exports.HTMLTextAreaElement = exports.HTMLTemplateElement = exports.HTMLStyleElement = exports.HTMLSlotElement = exports.HTMLScriptElement = exports.HTMLMetaElement = exports.HTMLMediaElement = exports.HTMLLinkElement = exports.HTMLLabelElement = exports.HTMLInputElement = exports.HTMLImageElement = exports.HTMLIFrameElement = exports.HTMLFormElement = exports.HTMLFormControlsCollection = exports.HTMLElement = exports.HTMLDocument = exports.HTMLDialogElement = exports.HTMLCollection = exports.HTMLBaseElement = exports.HTMLSelectElement = exports.HTMLUnknownElement = exports.HTMLOptionElement = exports.HTMLOptGroupElement = exports.HTMLAudioElement = exports.HTMLButtonElement = exports.HTMLAnchorElement = exports.GlobalWindow = exports.FormData = void 0;
exports.SVGSVGElement = exports.SVGGraphicsElement = exports.SVGElement = exports.SVGDocument = exports.Response = exports.ResizeObserver = exports.Request = exports.Range = exports.ProgressEvent = exports.ProcessingInstruction = exports.Permissions = exports.PermissionStatus = exports.NodeIterator = exports.NodeFilter = exports.Node = exports.MutationRecord = exports.MutationObserver = exports.MouseEvent = exports.Location = exports.KeyboardEvent = exports.InputEvent = exports.Image = exports.History = exports.Headers = exports.HTMLTrackElement = exports.HTMLParamElement = exports.HTMLObjectElement = exports.HTMLEmbedElement = exports.HTMLFrameSetElement = exports.HTMLFrameElement = exports.HTMLTableSectionElement = exports.HTMLTableRowElement = exports.HTMLTimeElement = exports.HTMLTableElement = exports.HTMLTableColElement = exports.HTMLTableCellElement = exports.HTMLTableCaptionElement = exports.HTMLSpanElement = exports.HTMLSourceElement = exports.HTMLQuoteElement = exports.HTMLProgressElement = exports.HTMLPictureElement = exports.HTMLOutputElement = exports.HTMLModElement = exports.HTMLMeterElement = exports.HTMLMarqueeElement = exports.HTMLMapElement = exports.HTMLLegendElement = exports.HTMLHtmlElement = exports.HTMLFontElement = void 0;
exports.XMLSerializer = exports.XMLParser = exports.XMLDocument = exports.Window = exports.WheelEvent = exports.VirtualConsolePrinter = exports.VirtualConsoleLogTypeEnum = exports.VirtualConsoleLogLevelEnum = exports.VirtualConsole = exports.URLSearchParams = exports.URL = exports.UIEvent = exports.TreeWalker = exports.Text = exports.SubmitEvent = exports.Storage = exports.ShadowRoot = exports.Selection = exports.Screen = void 0;
const url_1 = require("url");
Object.defineProperty(exports, "URLSearchParams", { enumerable: true, get: function () { return url_1.URLSearchParams; } });
const Browser_js_1 = __importDefault(require("./browser/Browser.cjs"));
exports.Browser = Browser_js_1.default;
const BrowserContext_js_1 = __importDefault(require("./browser/BrowserContext.cjs"));
exports.BrowserContext = BrowserContext_js_1.default;
const BrowserFrame_js_1 = __importDefault(require("./browser/BrowserFrame.cjs"));
exports.BrowserFrame = BrowserFrame_js_1.default;
const BrowserPage_js_1 = __importDefault(require("./browser/BrowserPage.cjs"));
exports.BrowserPage = BrowserPage_js_1.default;
const DetachedBrowser_js_1 = __importDefault(require("./browser/detached-browser/DetachedBrowser.cjs"));
exports.DetachedBrowser = DetachedBrowser_js_1.default;
const DetachedBrowserContext_js_1 = __importDefault(require("./browser/detached-browser/DetachedBrowserContext.cjs"));
exports.DetachedBrowserContext = DetachedBrowserContext_js_1.default;
const DetachedBrowserFrame_js_1 = __importDefault(require("./browser/detached-browser/DetachedBrowserFrame.cjs"));
exports.DetachedBrowserFrame = DetachedBrowserFrame_js_1.default;
const DetachedBrowserPage_js_1 = __importDefault(require("./browser/detached-browser/DetachedBrowserPage.cjs"));
exports.DetachedBrowserPage = DetachedBrowserPage_js_1.default;
const BrowserErrorCaptureEnum_js_1 = __importDefault(require("./browser/enums/BrowserErrorCaptureEnum.cjs"));
exports.BrowserErrorCaptureEnum = BrowserErrorCaptureEnum_js_1.default;
const BrowserNavigationCrossOriginPolicyEnum_js_1 = __importDefault(require("./browser/enums/BrowserNavigationCrossOriginPolicyEnum.cjs"));
exports.BrowserNavigationCrossOriginPolicyEnum = BrowserNavigationCrossOriginPolicyEnum_js_1.default;
const Clipboard_js_1 = __importDefault(require("./clipboard/Clipboard.cjs"));
exports.Clipboard = Clipboard_js_1.default;
const ClipboardItem_js_1 = __importDefault(require("./clipboard/ClipboardItem.cjs"));
exports.ClipboardItem = ClipboardItem_js_1.default;
const VirtualConsole_js_1 = __importDefault(require("./console/VirtualConsole.cjs"));
exports.VirtualConsole = VirtualConsole_js_1.default;
const VirtualConsolePrinter_js_1 = __importDefault(require("./console/VirtualConsolePrinter.cjs"));
exports.VirtualConsolePrinter = VirtualConsolePrinter_js_1.default;
const VirtualConsoleLogLevelEnum_js_1 = __importDefault(require("./console/enums/VirtualConsoleLogLevelEnum.cjs"));
exports.VirtualConsoleLogLevelEnum = VirtualConsoleLogLevelEnum_js_1.default;
const VirtualConsoleLogTypeEnum_js_1 = __importDefault(require("./console/enums/VirtualConsoleLogTypeEnum.cjs"));
exports.VirtualConsoleLogTypeEnum = VirtualConsoleLogTypeEnum_js_1.default;
const CSSRule_js_1 = __importDefault(require("./css/CSSRule.cjs"));
exports.CSSRule = CSSRule_js_1.default;
const CSSStyleSheet_js_1 = __importDefault(require("./css/CSSStyleSheet.cjs"));
exports.CSSStyleSheet = CSSStyleSheet_js_1.default;
const CSSStyleDeclaration_js_1 = __importDefault(require("./css/declaration/CSSStyleDeclaration.cjs"));
exports.CSSStyleDeclaration = CSSStyleDeclaration_js_1.default;
const CSSContainerRule_js_1 = __importDefault(require("./css/rules/CSSContainerRule.cjs"));
exports.CSSContainerRule = CSSContainerRule_js_1.default;
const CSSFontFaceRule_js_1 = __importDefault(require("./css/rules/CSSFontFaceRule.cjs"));
exports.CSSFontFaceRule = CSSFontFaceRule_js_1.default;
const CSSKeyframeRule_js_1 = __importDefault(require("./css/rules/CSSKeyframeRule.cjs"));
exports.CSSKeyframeRule = CSSKeyframeRule_js_1.default;
const CSSKeyframesRule_js_1 = __importDefault(require("./css/rules/CSSKeyframesRule.cjs"));
exports.CSSKeyframesRule = CSSKeyframesRule_js_1.default;
const CSSMediaRule_js_1 = __importDefault(require("./css/rules/CSSMediaRule.cjs"));
exports.CSSMediaRule = CSSMediaRule_js_1.default;
const CSSStyleRule_js_1 = __importDefault(require("./css/rules/CSSStyleRule.cjs"));
exports.CSSStyleRule = CSSStyleRule_js_1.default;
const CSSSupportsRule_js_1 = __importDefault(require("./css/rules/CSSSupportsRule.cjs"));
exports.CSSSupportsRule = CSSSupportsRule_js_1.default;
const CustomElementRegistry_js_1 = __importDefault(require("./custom-element/CustomElementRegistry.cjs"));
exports.CustomElementRegistry = CustomElementRegistry_js_1.default;
const DOMParser_js_1 = __importDefault(require("./dom-parser/DOMParser.cjs"));
exports.DOMParser = DOMParser_js_1.default;
const DataTransfer_js_1 = __importDefault(require("./event/DataTransfer.cjs"));
exports.DataTransfer = DataTransfer_js_1.default;
const DataTransferItem_js_1 = __importDefault(require("./event/DataTransferItem.cjs"));
exports.DataTransferItem = DataTransferItem_js_1.default;
const DataTransferItemList_js_1 = __importDefault(require("./event/DataTransferItemList.cjs"));
exports.DataTransferItemList = DataTransferItemList_js_1.default;
const Event_js_1 = __importDefault(require("./event/Event.cjs"));
exports.Event = Event_js_1.default;
const EventPhaseEnum_js_1 = __importDefault(require("./event/EventPhaseEnum.cjs"));
exports.EventPhaseEnum = EventPhaseEnum_js_1.default;
const EventTarget_js_1 = __importDefault(require("./event/EventTarget.cjs"));
exports.EventTarget = EventTarget_js_1.default;
const UIEvent_js_1 = __importDefault(require("./event/UIEvent.cjs"));
exports.UIEvent = UIEvent_js_1.default;
const AnimationEvent_js_1 = __importDefault(require("./event/events/AnimationEvent.cjs"));
exports.AnimationEvent = AnimationEvent_js_1.default;
const ClipboardEvent_js_1 = __importDefault(require("./event/events/ClipboardEvent.cjs"));
exports.ClipboardEvent = ClipboardEvent_js_1.default;
const CustomEvent_js_1 = __importDefault(require("./event/events/CustomEvent.cjs"));
exports.CustomEvent = CustomEvent_js_1.default;
const ErrorEvent_js_1 = __importDefault(require("./event/events/ErrorEvent.cjs"));
exports.ErrorEvent = ErrorEvent_js_1.default;
const FocusEvent_js_1 = __importDefault(require("./event/events/FocusEvent.cjs"));
exports.FocusEvent = FocusEvent_js_1.default;
const InputEvent_js_1 = __importDefault(require("./event/events/InputEvent.cjs"));
exports.InputEvent = InputEvent_js_1.default;
const KeyboardEvent_js_1 = __importDefault(require("./event/events/KeyboardEvent.cjs"));
exports.KeyboardEvent = KeyboardEvent_js_1.default;
const MouseEvent_js_1 = __importDefault(require("./event/events/MouseEvent.cjs"));
exports.MouseEvent = MouseEvent_js_1.default;
const ProgressEvent_js_1 = __importDefault(require("./event/events/ProgressEvent.cjs"));
exports.ProgressEvent = ProgressEvent_js_1.default;
const SubmitEvent_js_1 = __importDefault(require("./event/events/SubmitEvent.cjs"));
exports.SubmitEvent = SubmitEvent_js_1.default;
const WheelEvent_js_1 = __importDefault(require("./event/events/WheelEvent.cjs"));
exports.WheelEvent = WheelEvent_js_1.default;
const DOMException_js_1 = __importDefault(require("./exception/DOMException.cjs"));
exports.DOMException = DOMException_js_1.default;
const AbortController_js_1 = __importDefault(require("./fetch/AbortController.cjs"));
exports.AbortController = AbortController_js_1.default;
const AbortSignal_js_1 = __importDefault(require("./fetch/AbortSignal.cjs"));
exports.AbortSignal = AbortSignal_js_1.default;
const Headers_js_1 = __importDefault(require("./fetch/Headers.cjs"));
exports.Headers = Headers_js_1.default;
const Request_js_1 = __importDefault(require("./fetch/Request.cjs"));
exports.Request = Request_js_1.default;
const Response_js_1 = __importDefault(require("./fetch/Response.cjs"));
exports.Response = Response_js_1.default;
const Blob_js_1 = __importDefault(require("./file/Blob.cjs"));
exports.Blob = Blob_js_1.default;
const File_js_1 = __importDefault(require("./file/File.cjs"));
exports.File = File_js_1.default;
const FileReader_js_1 = __importDefault(require("./file/FileReader.cjs"));
exports.FileReader = FileReader_js_1.default;
const FormData_js_1 = __importDefault(require("./form-data/FormData.cjs"));
exports.FormData = FormData_js_1.default;
const History_js_1 = __importDefault(require("./history/History.cjs"));
exports.History = History_js_1.default;
const Location_js_1 = __importDefault(require("./url/Location.cjs"));
exports.Location = Location_js_1.default;
const MutationObserver_js_1 = __importDefault(require("./mutation-observer/MutationObserver.cjs"));
exports.MutationObserver = MutationObserver_js_1.default;
const MutationRecord_js_1 = __importDefault(require("./mutation-observer/MutationRecord.cjs"));
exports.MutationRecord = MutationRecord_js_1.default;
const Attr_js_1 = __importDefault(require("./nodes/attr/Attr.cjs"));
exports.Attr = Attr_js_1.default;
const Comment_js_1 = __importDefault(require("./nodes/comment/Comment.cjs"));
exports.Comment = Comment_js_1.default;
const DocumentFragment_js_1 = __importDefault(require("./nodes/document-fragment/DocumentFragment.cjs"));
exports.DocumentFragment = DocumentFragment_js_1.default;
const DocumentType_js_1 = __importDefault(require("./nodes/document-type/DocumentType.cjs"));
exports.DocumentType = DocumentType_js_1.default;
const Document_js_1 = __importDefault(require("./nodes/document/Document.cjs"));
exports.Document = Document_js_1.default;
const DOMRect_js_1 = __importDefault(require("./nodes/element/DOMRect.cjs"));
exports.DOMRect = DOMRect_js_1.default;
const Element_js_1 = __importDefault(require("./nodes/element/Element.cjs"));
exports.Element = Element_js_1.default;
const HTMLCollection_js_1 = __importDefault(require("./nodes/element/HTMLCollection.cjs"));
exports.HTMLCollection = HTMLCollection_js_1.default;
const HTMLAudioElement_js_1 = __importDefault(require("./nodes/html-audio-element/HTMLAudioElement.cjs"));
exports.HTMLAudioElement = HTMLAudioElement_js_1.default;
const HTMLBaseElement_js_1 = __importDefault(require("./nodes/html-base-element/HTMLBaseElement.cjs"));
exports.HTMLBaseElement = HTMLBaseElement_js_1.default;
const HTMLDialogElement_js_1 = __importDefault(require("./nodes/html-dialog-element/HTMLDialogElement.cjs"));
exports.HTMLDialogElement = HTMLDialogElement_js_1.default;
const HTMLDocument_js_1 = __importDefault(require("./nodes/html-document/HTMLDocument.cjs"));
exports.HTMLDocument = HTMLDocument_js_1.default;
const HTMLElement_js_1 = __importDefault(require("./nodes/html-element/HTMLElement.cjs"));
exports.HTMLElement = HTMLElement_js_1.default;
exports.HTMLTitleElement = HTMLElement_js_1.default;
exports.HTMLHeadElement = HTMLElement_js_1.default;
exports.HTMLBodyElement = HTMLElement_js_1.default;
exports.HTMLHeadingElement = HTMLElement_js_1.default;
exports.HTMLParagraphElement = HTMLElement_js_1.default;
exports.HTMLHRElement = HTMLElement_js_1.default;
exports.HTMLPreElement = HTMLElement_js_1.default;
exports.HTMLUListElement = HTMLElement_js_1.default;
exports.HTMLOListElement = HTMLElement_js_1.default;
exports.HTMLLIElement = HTMLElement_js_1.default;
exports.HTMLMenuElement = HTMLElement_js_1.default;
exports.HTMLDListElement = HTMLElement_js_1.default;
exports.HTMLDivElement = HTMLElement_js_1.default;
exports.HTMLAreaElement = HTMLElement_js_1.default;
exports.HTMLBRElement = HTMLElement_js_1.default;
exports.HTMLCanvasElement = HTMLElement_js_1.default;
exports.HTMLDataElement = HTMLElement_js_1.default;
exports.HTMLDataListElement = HTMLElement_js_1.default;
exports.HTMLDetailsElement = HTMLElement_js_1.default;
exports.HTMLDirectoryElement = HTMLElement_js_1.default;
exports.HTMLFieldSetElement = HTMLElement_js_1.default;
exports.HTMLFontElement = HTMLElement_js_1.default;
exports.HTMLHtmlElement = HTMLElement_js_1.default;
exports.HTMLLegendElement = HTMLElement_js_1.default;
exports.HTMLMapElement = HTMLElement_js_1.default;
exports.HTMLMarqueeElement = HTMLElement_js_1.default;
exports.HTMLMeterElement = HTMLElement_js_1.default;
exports.HTMLModElement = HTMLElement_js_1.default;
exports.HTMLOutputElement = HTMLElement_js_1.default;
exports.HTMLPictureElement = HTMLElement_js_1.default;
exports.HTMLProgressElement = HTMLElement_js_1.default;
exports.HTMLQuoteElement = HTMLElement_js_1.default;
exports.HTMLSourceElement = HTMLElement_js_1.default;
exports.HTMLSpanElement = HTMLElement_js_1.default;
exports.HTMLTableCaptionElement = HTMLElement_js_1.default;
exports.HTMLTableCellElement = HTMLElement_js_1.default;
exports.HTMLTableColElement = HTMLElement_js_1.default;
exports.HTMLTableElement = HTMLElement_js_1.default;
exports.HTMLTimeElement = HTMLElement_js_1.default;
exports.HTMLTableRowElement = HTMLElement_js_1.default;
exports.HTMLTableSectionElement = HTMLElement_js_1.default;
exports.HTMLFrameElement = HTMLElement_js_1.default;
exports.HTMLFrameSetElement = HTMLElement_js_1.default;
exports.HTMLEmbedElement = HTMLElement_js_1.default;
exports.HTMLObjectElement = HTMLElement_js_1.default;
exports.HTMLParamElement = HTMLElement_js_1.default;
exports.HTMLTrackElement = HTMLElement_js_1.default;
const HTMLFormControlsCollection_js_1 = __importDefault(require("./nodes/html-form-element/HTMLFormControlsCollection.cjs"));
exports.HTMLFormControlsCollection = HTMLFormControlsCollection_js_1.default;
const HTMLFormElement_js_1 = __importDefault(require("./nodes/html-form-element/HTMLFormElement.cjs"));
exports.HTMLFormElement = HTMLFormElement_js_1.default;
const HTMLIFrameElement_js_1 = __importDefault(require("./nodes/html-iframe-element/HTMLIFrameElement.cjs"));
exports.HTMLIFrameElement = HTMLIFrameElement_js_1.default;
const HTMLImageElement_js_1 = __importDefault(require("./nodes/html-image-element/HTMLImageElement.cjs"));
exports.HTMLImageElement = HTMLImageElement_js_1.default;
const Image_js_1 = __importDefault(require("./nodes/html-image-element/Image.cjs"));
exports.Image = Image_js_1.default;
const FileList_js_1 = __importDefault(require("./nodes/html-input-element/FileList.cjs"));
exports.FileList = FileList_js_1.default;
const HTMLInputElement_js_1 = __importDefault(require("./nodes/html-input-element/HTMLInputElement.cjs"));
exports.HTMLInputElement = HTMLInputElement_js_1.default;
const HTMLLabelElement_js_1 = __importDefault(require("./nodes/html-label-element/HTMLLabelElement.cjs"));
exports.HTMLLabelElement = HTMLLabelElement_js_1.default;
const HTMLLinkElement_js_1 = __importDefault(require("./nodes/html-link-element/HTMLLinkElement.cjs"));
exports.HTMLLinkElement = HTMLLinkElement_js_1.default;
const HTMLMediaElement_js_1 = __importDefault(require("./nodes/html-media-element/HTMLMediaElement.cjs"));
exports.HTMLMediaElement = HTMLMediaElement_js_1.default;
const HTMLMetaElement_js_1 = __importDefault(require("./nodes/html-meta-element/HTMLMetaElement.cjs"));
exports.HTMLMetaElement = HTMLMetaElement_js_1.default;
const HTMLScriptElement_js_1 = __importDefault(require("./nodes/html-script-element/HTMLScriptElement.cjs"));
exports.HTMLScriptElement = HTMLScriptElement_js_1.default;
const HTMLSlotElement_js_1 = __importDefault(require("./nodes/html-slot-element/HTMLSlotElement.cjs"));
exports.HTMLSlotElement = HTMLSlotElement_js_1.default;
const HTMLStyleElement_js_1 = __importDefault(require("./nodes/html-style-element/HTMLStyleElement.cjs"));
exports.HTMLStyleElement = HTMLStyleElement_js_1.default;
const HTMLTemplateElement_js_1 = __importDefault(require("./nodes/html-template-element/HTMLTemplateElement.cjs"));
exports.HTMLTemplateElement = HTMLTemplateElement_js_1.default;
const HTMLTextAreaElement_js_1 = __importDefault(require("./nodes/html-text-area-element/HTMLTextAreaElement.cjs"));
exports.HTMLTextAreaElement = HTMLTextAreaElement_js_1.default;
const HTMLVideoElement_js_1 = __importDefault(require("./nodes/html-video-element/HTMLVideoElement.cjs"));
exports.HTMLVideoElement = HTMLVideoElement_js_1.default;
const HTMLAnchorElement_js_1 = __importDefault(require("./nodes/html-anchor-element/HTMLAnchorElement.cjs"));
exports.HTMLAnchorElement = HTMLAnchorElement_js_1.default;
const HTMLButtonElement_js_1 = __importDefault(require("./nodes/html-button-element/HTMLButtonElement.cjs"));
exports.HTMLButtonElement = HTMLButtonElement_js_1.default;
const HTMLOptGroupElement_js_1 = __importDefault(require("./nodes/html-opt-group-element/HTMLOptGroupElement.cjs"));
exports.HTMLOptGroupElement = HTMLOptGroupElement_js_1.default;
const HTMLOptionElement_js_1 = __importDefault(require("./nodes/html-option-element/HTMLOptionElement.cjs"));
exports.HTMLOptionElement = HTMLOptionElement_js_1.default;
const HTMLSelectElement_js_1 = __importDefault(require("./nodes/html-select-element/HTMLSelectElement.cjs"));
exports.HTMLSelectElement = HTMLSelectElement_js_1.default;
const HTMLUnknownElement_js_1 = __importDefault(require("./nodes/html-unknown-element/HTMLUnknownElement.cjs"));
exports.HTMLUnknownElement = HTMLUnknownElement_js_1.default;
const Node_js_1 = __importDefault(require("./nodes/node/Node.cjs"));
exports.Node = Node_js_1.default;
const ProcessingInstruction_js_1 = __importDefault(require("./nodes/processing-instruction/ProcessingInstruction.cjs"));
exports.ProcessingInstruction = ProcessingInstruction_js_1.default;
const ShadowRoot_js_1 = __importDefault(require("./nodes/shadow-root/ShadowRoot.cjs"));
exports.ShadowRoot = ShadowRoot_js_1.default;
const SVGDocument_js_1 = __importDefault(require("./nodes/svg-document/SVGDocument.cjs"));
exports.SVGDocument = SVGDocument_js_1.default;
const SVGElement_js_1 = __importDefault(require("./nodes/svg-element/SVGElement.cjs"));
exports.SVGElement = SVGElement_js_1.default;
const SVGGraphicsElement_js_1 = __importDefault(require("./nodes/svg-element/SVGGraphicsElement.cjs"));
exports.SVGGraphicsElement = SVGGraphicsElement_js_1.default;
const SVGSVGElement_js_1 = __importDefault(require("./nodes/svg-element/SVGSVGElement.cjs"));
exports.SVGSVGElement = SVGSVGElement_js_1.default;
const Text_js_1 = __importDefault(require("./nodes/text/Text.cjs"));
exports.Text = Text_js_1.default;
const XMLDocument_js_1 = __importDefault(require("./nodes/xml-document/XMLDocument.cjs"));
exports.XMLDocument = XMLDocument_js_1.default;
const PermissionStatus_js_1 = __importDefault(require("./permissions/PermissionStatus.cjs"));
exports.PermissionStatus = PermissionStatus_js_1.default;
const Permissions_js_1 = __importDefault(require("./permissions/Permissions.cjs"));
exports.Permissions = Permissions_js_1.default;
const Range_js_1 = __importDefault(require("./range/Range.cjs"));
exports.Range = Range_js_1.default;
const ResizeObserver_js_1 = __importDefault(require("./resize-observer/ResizeObserver.cjs"));
exports.ResizeObserver = ResizeObserver_js_1.default;
const Screen_js_1 = __importDefault(require("./screen/Screen.cjs"));
exports.Screen = Screen_js_1.default;
const Selection_js_1 = __importDefault(require("./selection/Selection.cjs"));
exports.Selection = Selection_js_1.default;
const Storage_js_1 = __importDefault(require("./storage/Storage.cjs"));
exports.Storage = Storage_js_1.default;
const NodeFilter_js_1 = __importDefault(require("./tree-walker/NodeFilter.cjs"));
exports.NodeFilter = NodeFilter_js_1.default;
const NodeIterator_js_1 = __importDefault(require("./tree-walker/NodeIterator.cjs"));
exports.NodeIterator = NodeIterator_js_1.default;
const TreeWalker_js_1 = __importDefault(require("./tree-walker/TreeWalker.cjs"));
exports.TreeWalker = TreeWalker_js_1.default;
const URL_js_1 = __importDefault(require("./url/URL.cjs"));
exports.URL = URL_js_1.default;
const BrowserWindow_js_1 = __importDefault(require("./window/BrowserWindow.cjs"));
exports.BrowserWindow = BrowserWindow_js_1.default;
const GlobalWindow_js_1 = __importDefault(require("./window/GlobalWindow.cjs"));
exports.GlobalWindow = GlobalWindow_js_1.default;
const Window_js_1 = __importDefault(require("./window/Window.cjs"));
exports.Window = Window_js_1.default;
const XMLParser_js_1 = __importDefault(require("./xml-parser/XMLParser.cjs"));
exports.XMLParser = XMLParser_js_1.default;
const XMLSerializer_js_1 = __importDefault(require("./xml-serializer/XMLSerializer.cjs"));
exports.XMLSerializer = XMLSerializer_js_1.default;
