// happy-dom/cjs/PropertySymbol.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ownerElement = exports.ownerDocument = exports.observers = exports.observedAttributes = exports.observe = exports.nextActiveElement = exports.namedItems = exports.listeners = exports.listenerOptions = exports.isValue = exports.isValidPropertyName = exports.isInPassiveEventListener = exports.isFirstWriteAfterOpen = exports.isFirstWrite = exports.immediatePropagationStopped = exports.height = exports.happyDOMSettingsID = exports.getAttributeName = exports.formNode = exports.exceptionObserver = exports.evaluateScript = exports.evaluateCSS = exports.entries = exports.end = exports.dirtyness = exports.destroy = exports.defaultView = exports.data = exports.currentTarget = exports.currentScript = exports.cssText = exports.contentType = exports.contentLength = exports.connectToNode = exports.computedStyle = exports.classList = exports.children = exports.childNodes = exports.checked = exports.captureEventListenerCount = exports.callbacks = exports.cachedResponse = exports.cacheID = exports.buffer = exports.bodyBuffer = exports.asyncTaskManager = exports.appendNamedItem = exports.appendFormControlItem = exports.activeElement = exports.abort = void 0;
exports.offsetWidth = exports.offsetHeight = exports.isContentEditable = exports.contentEditable = exports.accessKeyLabel = exports.accessKey = exports.namespaceURI = exports.attributes = exports.scrollLeft = exports.scrollTop = exports.scrollWidth = exports.scrollHeight = exports.prefix = exports.tagName = exports.nodeType = exports.parentNode = exports.isConnected = exports.popup = exports.openerWindow = exports.openerFrame = exports.mutationObservers = exports.windowResizeListener = exports.window = exports.width = exports.value = exports.url = exports.updateOptionItems = exports.updateIndices = exports.unobserve = exports.textAreaNode = exports.target = exports.style = exports.start = exports.shadowRoot = exports.setupVMContext = exports.setNamedItemWithoutConsequences = exports.selection = exports.selectedness = exports.selectNode = exports.rootNode = exports.resetSelection = exports.removeNamedItemWithoutConsequences = exports.removeNamedItemIndex = exports.removeNamedItem = exports.removeFormControlItem = exports.relList = exports.registry = exports.referrer = exports.readyStateManager = exports.propagationStopped = void 0;
exports.host = exports.mode = exports.content = exports.options = exports.played = exports.seekable = exports.seeking = exports.videoTracks = exports.textTracks = exports.networkState = exports.ended = exports.error = exports.duration = exports.buffered = exports.preservesPitch = exports.defaultMuted = exports.muted = exports.defaultPlaybackRate = exports.playbackRate = exports.currentTime = exports.paused = exports.volume = exports.sheet = exports.files = exports.defaultChecked = exports.y = exports.x = exports.loading = exports.naturalWidth = exports.naturalHeight = exports.complete = exports.length = exports.elements = exports.returnValue = exports.validity = exports.validationMessage = exports.systemId = exports.publicId = exports.ownerWindow = exports.readyState = exports.implementation = exports.adoptedStyleSheets = exports.specified = exports.name = exports.clientTop = exports.clientLeft = exports.clientWidth = exports.clientHeight = exports.offsetTop = exports.offsetLeft = void 0;
exports.setURL = void 0;
exports.abort = Symbol('abort');
exports.activeElement = Symbol('activeElement');
exports.appendFormControlItem = Symbol('appendFormControlItem');
exports.appendNamedItem = Symbol('appendNamedItem');
exports.asyncTaskManager = Symbol('asyncTaskManager');
exports.bodyBuffer = Symbol('bodyBuffer');
exports.buffer = Symbol('buffer');
exports.cacheID = Symbol('cacheID');
exports.cachedResponse = Symbol('cachedResponse');
exports.callbacks = Symbol('callbacks');
exports.captureEventListenerCount = Symbol('captureEventListenerCount');
exports.checked = Symbol('checked');
exports.childNodes = Symbol('childNodes');
exports.children = Symbol('children');
exports.classList = Symbol('classList');
exports.computedStyle = Symbol('computedStyle');
exports.connectToNode = Symbol('connectToNode');
exports.contentLength = Symbol('contentLength');
exports.contentType = Symbol('contentType');
exports.cssText = Symbol('cssText');
exports.currentScript = Symbol('currentScript');
exports.currentTarget = Symbol('currentTarget');
exports.data = Symbol('data');
exports.defaultView = Symbol('defaultView');
exports.destroy = Symbol('destroy');
exports.dirtyness = Symbol('dirtyness');
exports.end = Symbol('end');
exports.entries = Symbol('entries');
exports.evaluateCSS = Symbol('evaluateCSS');
exports.evaluateScript = Symbol('evaluateScript');
exports.exceptionObserver = Symbol('exceptionObserver');
exports.formNode = Symbol('formNode');
exports.getAttributeName = Symbol('getAttributeName');
exports.happyDOMSettingsID = Symbol('happyDOMSettingsID');
exports.height = Symbol('height');
exports.immediatePropagationStopped = Symbol('immediatePropagationStopped');
exports.isFirstWrite = Symbol('isFirstWrite');
exports.isFirstWriteAfterOpen = Symbol('isFirstWriteAfterOpen');
exports.isInPassiveEventListener = Symbol('isInPassiveEventListener');
exports.isValidPropertyName = Symbol('isValidPropertyName');
exports.isValue = Symbol('isValue');
exports.listenerOptions = Symbol('listenerOptions');
exports.listeners = Symbol('listeners');
exports.namedItems = Symbol('namedItems');
exports.nextActiveElement = Symbol('nextActiveElement');
exports.observe = Symbol('observe');
exports.observedAttributes = Symbol('observedAttributes');
exports.observers = Symbol('observers');
exports.ownerDocument = Symbol('ownerDocument');
exports.ownerElement = Symbol('ownerElement');
exports.propagationStopped = Symbol('propagationStopped');
exports.readyStateManager = Symbol('readyStateManager');
exports.referrer = Symbol('referrer');
exports.registry = Symbol('registry');
exports.relList = Symbol('relList');
exports.removeFormControlItem = Symbol('removeFormControlItem');
exports.removeNamedItem = Symbol('removeNamedItem');
exports.removeNamedItemIndex = Symbol('removeNamedItemIndex');
exports.removeNamedItemWithoutConsequences = Symbol('removeNamedItemWithoutConsequences');
exports.resetSelection = Symbol('resetSelection');
exports.rootNode = Symbol('rootNode');
exports.selectNode = Symbol('selectNode');
exports.selectedness = Symbol('selectedness');
exports.selection = Symbol('selection');
exports.setNamedItemWithoutConsequences = Symbol('setNamedItemWithoutConsequences');
exports.setupVMContext = Symbol('setupVMContext');
exports.shadowRoot = Symbol('shadowRoot');
exports.start = Symbol('start');
exports.style = Symbol('style');
exports.target = Symbol('target');
exports.textAreaNode = Symbol('textAreaNode');
exports.unobserve = Symbol('unobserve');
exports.updateIndices = Symbol('updateIndices');
exports.updateOptionItems = Symbol('updateOptionItems');
exports.url = Symbol('url');
exports.value = Symbol('value');
exports.width = Symbol('width');
exports.window = Symbol('window');
exports.windowResizeListener = Symbol('windowResizeListener');
exports.mutationObservers = Symbol('mutationObservers');
exports.openerFrame = Symbol('openerFrame');
exports.openerWindow = Symbol('openerFrame');
exports.popup = Symbol('popup');
exports.isConnected = Symbol('isConnected');
exports.parentNode = Symbol('parentNode');
exports.nodeType = Symbol('nodeType');
exports.tagName = Symbol('tagName');
exports.prefix = Symbol('prefix');
exports.scrollHeight = Symbol('scrollHeight');
exports.scrollWidth = Symbol('scrollWidth');
exports.scrollTop = Symbol('scrollTop');
exports.scrollLeft = Symbol('scrollLeft');
exports.attributes = Symbol('attributes');
exports.namespaceURI = Symbol('namespaceURI');
exports.accessKey = Symbol('accessKey');
exports.accessKeyLabel = Symbol('accessKeyLabel');
exports.contentEditable = Symbol('contentEditable');
exports.isContentEditable = Symbol('isContentEditable');
exports.offsetHeight = Symbol('offsetHeight');
exports.offsetWidth = Symbol('offsetWidth');
exports.offsetLeft = Symbol('offsetLeft');
exports.offsetTop = Symbol('offsetTop');
exports.clientHeight = Symbol('clientHeight');
exports.clientWidth = Symbol('clientWidth');
exports.clientLeft = Symbol('clientLeft');
exports.clientTop = Symbol('clientTop');
exports.name = Symbol('name');
exports.specified = Symbol('specified');
exports.adoptedStyleSheets = Symbol('adoptedStyleSheets');
exports.implementation = Symbol('implementation');
exports.readyState = Symbol('readyState');
exports.ownerWindow = Symbol('ownerWindow');
exports.publicId = Symbol('publicId');
exports.systemId = Symbol('systemId');
exports.validationMessage = Symbol('validationMessage');
exports.validity = Symbol('validity');
exports.returnValue = Symbol('returnValue');
exports.elements = Symbol('elements');
exports.length = Symbol('length');
exports.complete = Symbol('complete');
exports.naturalHeight = Symbol('naturalHeight');
exports.naturalWidth = Symbol('naturalWidth');
exports.loading = Symbol('loading');
exports.x = Symbol('x');
exports.y = Symbol('y');
exports.defaultChecked = Symbol('defaultChecked');
exports.files = Symbol('files');
exports.sheet = Symbol('sheet');
exports.volume = Symbol('volume');
exports.paused = Symbol('paused');
exports.currentTime = Symbol('currentTime');
exports.playbackRate = Symbol('playbackRate');
exports.defaultPlaybackRate = Symbol('defaultPlaybackRate');
exports.muted = Symbol('muted');
exports.defaultMuted = Symbol('defaultMuted');
exports.preservesPitch = Symbol('preservesPitch');
exports.buffered = Symbol('buffered');
exports.duration = Symbol('duration');
exports.error = Symbol('error');
exports.ended = Symbol('ended');
exports.networkState = Symbol('networkState');
exports.textTracks = Symbol('textTracks');
exports.videoTracks = Symbol('videoTracks');
exports.seeking = Symbol('seeking');
exports.seekable = Symbol('seekable');
exports.played = Symbol('played');
exports.options = Symbol('options');
exports.content = Symbol('content');
exports.mode = Symbol('mode');
exports.host = Symbol('host');
exports.setURL = Symbol('setURL');
